src/core/queryObserver.ts

```diff
@@ -763,6 +763,7 @@ function shouldFetchOptionally(
   return (
     options.enabled !== false &&
     (query !== prevQuery || prevOptions.enabled === false) &&
+    query.state.status !== 'error' &&
```

Comment:
do you know which implications this can have on non-suspense / errorBoundary usages?

Reply:
@TkDodo i didn't catch (

so, how about using `options.suspense` condition?

Reply:
> what's the problem?

I was just trying to understand `shouldFetchOptionally` and if there is a scenario where we are in error state, but still want to fetch, and I think I just found something. Consider the scenario:

- you have an `enabled` query that errors
- you disable that query
- you enable it again
--> we should now refetch, but we don't

[Here](https://codesandbox.io/s/error-fetch-on-mount-39wc8?file=/src/App.tsx) is a working codesandbox for that scenario. You can see in the console that the `Error` is logged every time we fetch (clicking the button twice for disable/enable), and [here](https://codesandbox.io/s/error-fetch-on-mount-forked-6q6ug?file=/src/App.tsx) is the same example forked with the codesanbox preview build of this PR. You can see that we never fetch if we enable the query again :/

---

your commit message suggests we want to avoid fetching  if the status changes from `loading` to `error`, maybe we need to check the status of the previousQuery as well?

Reply:
Thanks for detail! i understood.

> we need to check the status of the previousQuery as well?

sounds good!

```
prevQuery.state.status !== 'error'
```
also working!

Reply:
> You can see that we never fetch if we enable the query again

i see. i'll check

Reply:
how about this?

```
(prevQuery.state.status !== 'error' || prevOptions.enabled !== options.enabled) &&
```

Reply:
Is this approach correct?

```ts
(prevQuery.state.status !== 'error' ||
      prevOptions.enabled !== options.enabled) &&
```

🤔

Reply:
cool, thanks for the additional test. I think that should work 👍